### ISSUE TITLE
Incremental parsing and language server protocol

### ISSUE THREAD

**Issue by Pyromuffin** [tree-sitter/tree-sitter#613]
Hi, I'm implementing a language server for a new language, and I'm using tree sitter. It's been pretty great so far.

However, I have a few issues:
Does it seem reasonable to parse a 500 line document in about 5 ms? (this seems extremely slow to me). Because of this slowness, I've been implementing incremental parsing, and I'm having some trouble with this.

1) TSInputEdit wants both line/column and byte offset, and the language server protocol only gives line/col information, so to efficiently get this byte offset information i have to implement some kind of acceleration structure, duplicating the work of the text editor.

2) Ignoring points, I don't understand the difference between new_end_byte and old_end_byte, and I can't find any examples of how to use this function with a corresponding code string, the edited string, and their expected s-expressions. What do I pass in for an insert, a replacement, or a delete? The tests in tree_test.rs don't indicate what you should pass to ts_parser_parse, or what the edited string looks like.

3) Do i call ts_parser_parse_string() with just the edit, or the entire edited document. My tests seem to indicate that it wants the entire edited document, which is problematic with language server protocol, because you are given two document synchronization modes: full and incremental. In full sync mode, you get the whole document each time it changes, and it *doesn't* tell where the edits are. In incremental mode, you get only the changes, their locations (in row/col). This means that to give the edited document to ts_parser_parse_string, i would then have to efficiently apply the edit to the document - replicating the work of the text editor.

Basically, my issue is that to use incremental parsing with LSP, I need to implement a text editor backend, which seems like a bit of a waste. 

I am also open to the possibility that I've written my grammar inefficiently, and 5 ms for 500 lines (with optimized code) is due to my inexperience writing parser grammars.

Thanks,
Kelly

**Comment by Aerijo** [comment]
I don't know if 5ms is reasonable or not, but Tree-sitter is designed for incremental parsing, so being slower the first time than a non-incremental parser would be expected. The benefit is seen on the subsequent parses.

Also it doesn't have any opinions about how you are storing your text; the [providing your code](https://tree-sitter.github.io/tree-sitter/using-parsers#providing-the-code) sections shows how a generic `read` function can be provided to walk whatever data structure you are storing the text in. Tree-sitter doesn't store any of the contents itself; AFAIK, the TS types only store information in terms of byte offsets and row/column. If you want to find out what text is in the given range, you'd need to look this up in your text buffer.

So you'd only really see a benefit with LSP when using incremental document updates, and using these to then do an incremental parsing update. The general flow would be LSP update -> update your text buffer -> update your parse tree. 

You do need to replicate the work of the text editor, but the only way not to would be to share the text editor's buffer directly, which doesn't work with LSP. You don't need to make it complicated though; a simple string or array of lines should be more than enough to get started. I wouldn't worry about representing enormous files until your server works at all, or you can find an existing text buffer library.

I can't help you on the other points, or how to do the edit though. You may find you need to run through the edited text to find it's end point (row & column).

**Comment by Pyromuffin** [comment]
Thanks Aerijo, I've already got my LSP working with semantic tokens and code completion, and I'm trying to improve the speed, and my profiling shows the majority of the time is spent in ts_parser_parse_string.

If TS isn't storing the buffer, then yeah, it would need the entire document each time. I've already implemented array of strings for accelerating the byte offset lookup work, so I'll see if i can modify that to efficiently apply edits.

However, it would still be quite useful to me to see some examples of ts_edit_tree()

**Comment by maxbrunsfeld** [comment]
> it would still be quite useful to me to see some examples of ts_edit_tree()

I'm not sure what language you're using, so some of these examples may be more helpful than others. 

* The README for the rust binding has an example: https://github.com/tree-sitter/tree-sitter/tree/master/lib/binding_rust#editing.
* Many of the Rust unit tests ([example](https://github.com/tree-sitter/tree-sitter/blob/a04d688d3895e17bf6111722aa57b7148d8f9df1/cli/src/tests/parser_test.rs#L305)) use [this helper function](https://github.com/tree-sitter/tree-sitter/blob/a04d688d3895e17bf6111722aa57b7148d8f9df1/cli/src/parse.rs#L206) for dealing with edits
* The README for the JavaScript binding [has another example](https://github.com/tree-sitter/tree-sitter/blob/a04d688d3895e17bf6111722aa57b7148d8f9df1/lib/binding_web/README.md#editing)
* The [JavaScript unit tests ](https://github.com/tree-sitter/tree-sitter/blob/a04d688d3895e17bf6111722aa57b7148d8f9df1/lib/binding_web/test/tree-test.js#L23) use a [similar helper function](https://github.com/tree-sitter/tree-sitter/blob/a04d688d3895e17bf6111722aa57b7148d8f9df1/lib/binding_web/test/tree-test.js#L328)

The idea is that `start` and `old_end` describe the range of text that was removed, and `start` and `new_end` describe the new range of text that was inserted. So for insertions, `old_end == start`, and for deletions, `new_end == start`.

> so to efficiently get this byte offset information i have to implement some kind of acceleration structure, duplicating the work of the text editor.

I think the duplication that you're noticing is a property of *any* LSP server; it's not specifically related to Tree-sitter. Fundamentally, with LSP, you must maintaining a replica of the Text editor's documents in a separate process. We could have designed Tree-sitter to store this data for you, but that would be very wasteful in other contexts, because in many use applications, you *already* need to store your own copy of the text for other purposes.

I'm going to close this out, but feel free to reopen if you are still having problems using `ts_tree_edit`.
